etcdserver/v3_server.go

```diff
@@ -641,3 +646,89 @@ func (s *EtcdServer) processInternalRaftRequest(ctx context.Context, r pb.Intern
 
 // Watchable returns a watchable interface attached to the etcdserver.
 func (s *EtcdServer) Watchable() mvcc.WatchableKV { return s.KV() }
+
+func (s *EtcdServer) linearizableReadLoop() {
+	var (
+		rs       raft.ReadState
+		requests int
+	)
+	ctx := make([]byte, 8)
+	s.readNotifyc = make(chan struct{})
+
+	for {
+		binary.BigEndian.PutUint64(ctx, s.reqIDGen.Next())
+
+		for (len(s.readwaitc) > 0 && requests < 256) || (len(s.readwaitc) == 0 && requests == 0) {
```

Comment:
Both latency and throughput. 

A simple limiter wont work well either. For example, we set limiter to 10/seconds. Then if we have N read requests from one client coming in a queue, then avg latency is 100ms. 

The current batching is more like the mechanism described here:
http://mechanical-sympathy.blogspot.tw/2011/10/smart-batching.html